File: discord/jda/src/main/java/dev/triumphteam/cmd/jda/SlashCommandManager.java
/**
 * MIT License
 *
 * Copyright (c) 2019-2021 Matt
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package dev.triumphteam.cmd.jda;

import dev.triumphteam.cmd.core.CommandManager;
import dev.triumphteam.cmd.core.argument.InternalArgument;
import dev.triumphteam.cmd.core.command.ArgumentInput;
import dev.triumphteam.cmd.core.command.InternalRootCommand;
import dev.triumphteam.cmd.core.exceptions.CommandExecutionException;
import dev.triumphteam.cmd.core.extension.registry.MessageRegistry;
import dev.triumphteam.cmd.core.extension.sender.SenderExtension;
import dev.triumphteam.cmd.core.message.MessageKey;
import dev.triumphteam.cmd.core.processor.RootCommandProcessor;
import dev.triumphteam.cmd.discord.LeafResult;
import dev.triumphteam.cmd.discord.ProvidedInternalArgument;
import dev.triumphteam.cmd.discord.choices.ChoiceKey;
import dev.triumphteam.cmd.jda.sender.SlashSender;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.IMentionable;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.Role;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel;
import net.dv8tion.jda.api.entities.channel.middleman.GuildChannel;
import net.dv8tion.jda.api.entities.channel.middleman.MessageChannel;
import net.dv8tion.jda.api.events.interaction.command.CommandAutoCompleteInteractionEvent;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.interactions.AutoCompleteQuery;
import net.dv8tion.jda.api.interactions.commands.CommandInteractionPayload;
import net.dv8tion.jda.api.interactions.commands.OptionMapping;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.reflect.InvocationTargetException;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Collections;
import java.util.Deque;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;
import java.util.function.Supplier;
import java.util.stream.Collectors;

import static dev.triumphteam.cmd.discord.CommandWalkUtil.findExecutable;

/**
 * Command Manager for Slash Commands.
 * Allows for registering of global and guild-specific commands.
 * As well as the implementation of custom command senders.
 *
 * @param <S> The sender type.
 */
public final class SlashCommandManager<S> extends CommandManager<SlashSender, S, SlashCommandOptions<S>> {

    private final JDA jda;

    private final SlashRegistryContainer<S> registryContainer;

    private final Map<String, InternalRootCommand<SlashSender, S>> globalCommands = new HashMap<>();
    private final Map<Long, Map<String, InternalRootCommand<SlashSender, S>>> guildCommands = new HashMap<>();

    private SlashCommandManager(
            final @NotNull JDA jda,
            final @NotNull SlashCommandOptions<S> commandOptions,
            final @NotNull SlashRegistryContainer<S> registryContainer
    ) {
        super(commandOptions);
        this.jda = jda;
        this.registryContainer = registryContainer;

        // All use the same factory
        final InternalArgument.Factory<S> jdaArgumentFactory = ProvidedInternalArgument::new;

        registerArgument(User.class, ProvidedUserInternalArgument::new);
        registerArgument(Role.class, jdaArgumentFactory);
        registerArgument(Member.class, jdaArgumentFactory);
        registerArgument(GuildChannel.class, jdaArgumentFactory);
        registerArgument(TextChannel.class, jdaArgumentFactory);
        registerArgument(MessageChannel.class, jdaArgumentFactory);
        registerArgument(Message.Attachment.class, jdaArgumentFactory);
        registerArgument(IMentionable.class, jdaArgumentFactory);

        if (commandOptions.autoRegisterListener()) {
            try {
                jda.awaitReady();
                jda.addEventListener(new SlashCommandsListener(this));
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }
    }

    @Contract("_, _, _ -> new")
    public static <S> @NotNull SlashCommandManager<S> create(
            final @NotNull JDA jda,
            final @NotNull SenderExtension<SlashSender, S> senderExtension,
            final @NotNull Consumer<SlashCommandOptions.Builder<S>> builder
    ) {
        final SlashRegistryContainer<S> registryContainer = new SlashRegistryContainer<>();
        final SlashCommandOptions.Builder<S> extensionBuilder = new SlashCommandOptions.Builder<>(registryContainer);
        builder.accept(extensionBuilder);
        return new SlashCommandManager<>(jda, extensionBuilder.build(senderExtension), registryContainer);
    }

    @Contract("_, _ -> new")
    public static @NotNull SlashCommandManager<SlashSender> create(
            final @NotNull JDA jda,
            final @NotNull Consumer<SlashCommandOptions.Builder<SlashSender>> builder
    ) {
        final SlashRegistryContainer<SlashSender> registryContainer = new SlashRegistryContainer<>();
        final SlashCommandOptions.Builder<SlashSender> extensionBuilder = new SlashCommandOptions.Builder<>(registryContainer);

        // Setup defaults for Bukkit
        final MessageRegistry<SlashSender> messageRegistry = registryContainer.getMessageRegistry();
        setUpDefaults(messageRegistry);

        // Then accept configured values
        builder.accept(extensionBuilder);
        return new SlashCommandManager<>(jda, extensionBuilder.build(new SlashSenderExtension()), registryContainer);
    }

    @Contract("_ -> new")
    public static @NotNull SlashCommandManager<SlashSender> create(final @NotNull JDA jda) {
        return create(jda, builder -> {});
    }

    private static void setUpDefaults(final @NotNull MessageRegistry<SlashSender> registry) {
        registry.register(MessageKey.UNKNOWN_COMMAND, (sender, context) -> sender.reply("Unknown command: `" + context.getInvalidInput() + "`.").setEphemeral(true).queue());
        registry.register(MessageKey.TOO_MANY_ARGUMENTS, (sender, context) -> sender.reply("Invalid usage.").setEphemeral(true).queue());
        registry.register(MessageKey.NOT_ENOUGH_ARGUMENTS, (sender, context) -> sender.reply("Invalid usage.").setEphemeral(true).queue());
        registry.register(MessageKey.INVALID_ARGUMENT, (sender, context) -> sender.reply("Invalid argument `" + context.getInvalidInput() + "` for type `" + context.getArgumentType().getSimpleName() + "`.").setEphemeral(true).queue());
    }

    public void registerChoices(final @NotNull ChoiceKey key, final @NotNull Supplier<List<String>> choiceSupplier) {
        registryContainer.getChoiceRegistry().register(key, choiceSupplier);
    }

    public void execute(final @NotNull SlashCommandInteractionEvent event) {
        final Deque<String> commands = new ArrayDeque<>(Arrays.asList(event.getFullCommandName().split(" ")));

        final SenderExtension<SlashSender, S> senderExtension = getCommandOptions().getCommandExtensions().getSenderExtension();
        final S sender = senderExtension.map(new InteractionCommandSender(event));

        final LeafResult<SlashSender, S> result = findExecutable(sender, getAppropriateMap(event), commands, true);
        if (result == null) return;

        // Mapping of arguments
        final Map<String, ArgumentInput> arguments = new HashMap<>();
        for (final OptionMapping option : event.getOptions()) {
            arguments.put(option.getName(), JdaMappingUtil.parsedValueFromType(option));
        }

        try {
            result.getCommand().execute(sender, result.getInstanceSupplier(), arguments);
        } catch (final @NotNull Throwable exception) {
            throw new CommandExecutionException("An error occurred while executing the command")
                    .initCause(exception instanceof InvocationTargetException ? exception.getCause() : exception);
        }
    }

    public void suggest(final @NotNull CommandAutoCompleteInteractionEvent event) {
        final Deque<String> commands = new ArrayDeque<>(Arrays.asList(event.getFullCommandName().split(" ")));

        final SenderExtension<SlashSender, S> senderExtension = getCommandOptions().getCommandExtensions().getSenderExtension();
        final S sender = senderExtension.map(new SuggestionCommandSender(event));

        final LeafResult<SlashSender, S> result = findExecutable(sender, getAppropriateMap(event), commands, true);
        if (result == null) return;

        final AutoCompleteQuery option = event.getFocusedOption();
        final List<String> arguments = event.getOptions().stream().map(OptionMapping::getAsString).collect(Collectors.toList());

        // On discord platforms, we don't need to navigate the command and can go straight into the argument.
        final InternalArgument<S> argument = result.getCommand().getArgument(option.getName());
        if (argument == null) return;

        final List<String> suggestions = argument.suggestions(sender, option.getValue(), arguments)
                .stream()
                .limit(25) // Discord only handles 25 at a time, :pensive:.
                .collect(Collectors.toList());

        event.replyChoices(JdaMappingUtil.mapChoices(suggestions, JdaMappingUtil.fromType(argument.getType()))).queue();
    }

    @Override
    public void registerCommand(final @NotNull Object command) {
        final RootCommandProcessor<SlashSender, S> processor = new RootCommandProcessor<>(
                command,
                getRegistryContainer(),
                getCommandOptions()
        );

        final String name = processor.getName();

        // Get or add a command, then add its sub commands
        final InternalRootCommand<SlashSender, S> rootCommand = globalCommands
                .computeIfAbsent(name, it -> new InternalRootCommand<>(processor));

        rootCommand.addCommands(command, processor.commands(rootCommand));
    }

    public void registerCommand(final @NotNull Guild guild, final @NotNull List<@NotNull Object> commands) {
        for (final Object command : commands) {
            registerCommand(guild.getIdLong(), command);
        }
    }

    public void registerCommand(final @NotNull Guild guild, final @NotNull Object @NotNull ... commands) {
        for (final Object command : commands) {
            registerCommand(guild.getIdLong(), command);
        }
    }

    public void registerCommand(final @NotNull Guild guild, final @NotNull Object command) {
        registerCommand(guild.getIdLong(), command);
    }

    public void registerCommand(final @NotNull Long guildId, final @NotNull List<@NotNull Object> commands) {
        for (final Object command : commands) {
            registerCommand(guildId, command);
        }
    }

    public void registerCommand(final @NotNull Long guildId, final @NotNull Object @NotNull ... commands) {
        for (final Object command : commands) {
            registerCommand(guildId, command);
        }
    }

    public void registerCommand(final @NotNull Long guildId, final @NotNull Object command) {
        final RootCommandProcessor<SlashSender, S> processor = new RootCommandProcessor<>(
                command,
                getRegistryContainer(),
                getCommandOptions()
        );

        final String name = processor.getName();

        // Get or add a command, then add its sub commands
        final InternalRootCommand<SlashSender, S> rootCommand = guildCommands
                .computeIfAbsent(guildId, it -> new HashMap<>())
                .computeIfAbsent(name, it -> new InternalRootCommand<>(processor));

        rootCommand.addCommands(command, processor.commands(rootCommand));
    }

    public void pushGuildCommands() {
        guildCommands.forEach((key, value) -> {
            final Guild guild = jda.getGuildById(key);
            if (guild == null) return;
            guild.updateCommands()
                    .addCommands(value.values().stream().map(JdaMappingUtil::mapCommand).collect(Collectors.toList()))
                    .queue();
        });
    }

    public void pushGlobalCommands() {
        jda.updateCommands()
                .addCommands(globalCommands.values().stream().map(JdaMappingUtil::mapCommand).collect(Collectors.toList()))
                .queue();
    }

    public void pushCommands() {
        pushGlobalCommands();
        pushGuildCommands();
    }

    @Override
    public void unregisterCommand(final @NotNull Object command) {

    }

    @Override
    protected @NotNull SlashRegistryContainer<S> getRegistryContainer() {
        return registryContainer;
    }

    private @NotNull Map<String, InternalRootCommand<SlashSender, S>> getAppropriateMap(final @NotNull CommandInteractionPayload event) {
        if (event.isGlobalCommand()) return globalCommands;

        final Guild guild = event.getGuild();
        if (guild == null) return Collections.emptyMap();
        return guildCommands.getOrDefault(guild.getIdLong(), Collections.emptyMap());
    }

    private @Nullable InternalRootCommand<SlashSender, S> getCommand(
            final @NotNull CommandInteractionPayload event,
            final @NotNull String name
    ) {
        if (event.isGlobalCommand()) return globalCommands.get(name);

        final Guild guild = event.getGuild();
        // Should never be null here
        if (guild == null) return null;
        return guildCommands
                .getOrDefault(guild.getIdLong(), Collections.emptyMap())
                .get(name);
    }
}

File: discord/kord/slash/src/main/kotlin/dev/triumphteam/cmds/kord/SlashCommandManager.kt
/**
 * MIT License
 *
 * Copyright (c) 2019-2021 Matt
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package dev.triumphteam.cmds.kord

import dev.kord.common.entity.Snowflake
import dev.kord.core.Kord
import dev.kord.core.behavior.interaction.suggestString
import dev.kord.core.entity.Attachment
import dev.kord.core.entity.Entity
import dev.kord.core.entity.Guild
import dev.kord.core.entity.Member
import dev.kord.core.entity.Role
import dev.kord.core.entity.User
import dev.kord.core.entity.channel.GuildChannel
import dev.kord.core.entity.channel.MessageChannel
import dev.kord.core.entity.channel.TextChannel
import dev.kord.core.entity.interaction.GroupCommand
import dev.kord.core.entity.interaction.ResolvableOptionValue
import dev.kord.core.event.gateway.ReadyEvent
import dev.kord.core.event.interaction.ChatInputCommandInteractionCreateEvent
import dev.kord.core.event.interaction.GuildAutoCompleteInteractionCreateEvent
import dev.kord.core.on
import dev.kord.rest.builder.interaction.group
import dev.kord.rest.builder.interaction.subCommand
import dev.triumphteam.cmd.core.CommandManager
import dev.triumphteam.cmd.core.argument.InternalArgument
import dev.triumphteam.cmd.core.command.ParentSubCommand
import dev.triumphteam.cmd.core.command.RootCommand
import dev.triumphteam.cmd.core.command.SubCommand
import dev.triumphteam.cmd.core.extension.registry.RegistryContainer
import dev.triumphteam.cmd.core.extension.sender.SenderExtension
import dev.triumphteam.cmd.core.processor.RootCommandProcessor
import dev.triumphteam.cmd.core.util.Pair
import dev.triumphteam.cmd.discord.ProvidedInternalArgument
import dev.triumphteam.cmds.kord.sender.SlashSender
import kotlinx.coroutines.launch
import java.util.ArrayDeque

public fun SlashCommandManager(
    kord: Kord,
    builder: SlashCommandOptions.Builder<SlashSender>.() -> Unit = {},
): SlashCommandManager<SlashSender> = SlashCommandManager(kord, SlashSenderExtension(), builder)

public fun <S> SlashCommandManager(
    kord: Kord,
    senderExtension: SenderExtension<SlashSender, S>,
    builder: SlashCommandOptions.Builder<S>.() -> Unit = {},
): SlashCommandManager<S> {
    val registryContainer = SlashRegistryContainer<S>()
    return SlashCommandManager(
        kord,
        SlashCommandOptions
            .Builder(registryContainer, kord)
            .apply(builder)
            .build(senderExtension),
        registryContainer
    )
}

public class SlashCommandManager<S>(
    private val kord: Kord,
    commandOptions: SlashCommandOptions<S>,
    private val registryContainer: SlashRegistryContainer<S>,
) : CommandManager<SlashSender, S, SlashCommandOptions<S>>(commandOptions) {

    private val globalCommands: MutableMap<String, RootCommand<SlashSender, S>> = mutableMapOf()
    private val guildCommands: MutableMap<Snowflake, MutableMap<String, RootCommand<SlashSender, S>>> = mutableMapOf()

    private val commandQueue: MutableList<suspend () -> Unit> = mutableListOf()

    private var isKordReady = false

    init {
        kord.on(consumer = ::execute)
        kord.on(consumer = ::suggest)
        kord.on<ReadyEvent> {
            // Sets kord to ready
            isKordReady = true
            commandQueue.forEach { it.invoke() }
        }

        // All use the same factory
        val jdaArgumentFactory = InternalArgument.Factory<S> { meta, name, description, type, suggestion, optional ->
            ProvidedInternalArgument(meta, name, description, type, suggestion, optional)
        }

        registerArgument(User::class.java, jdaArgumentFactory)
        registerArgument(Role::class.java, jdaArgumentFactory)
        registerArgument(Member::class.java, jdaArgumentFactory)
        registerArgument(GuildChannel::class.java, jdaArgumentFactory)
        registerArgument(TextChannel::class.java, jdaArgumentFactory)
        registerArgument(MessageChannel::class.java, jdaArgumentFactory)
        registerArgument(Attachment::class.java, jdaArgumentFactory)
        registerArgument(Entity::class.java, jdaArgumentFactory)
    }

    override fun registerCommand(command: Any) {
        TODO("Not yet implemented")
    }

    public fun registerCommand(guild: Guild, commands: List<Any>): Unit = commands.forEach {
        registerCommand(guild, it)
    }

    public fun registerCommand(guild: Guild, vararg commands: Any): Unit = commands.forEach {
        registerCommand(guild, it)
    }

    public fun registerCommand(guild: Guild, command: Any) {
        registerCommand(guild.id, command)
    }

    public fun registerCommand(guildId: Snowflake, command: Any) {
        val processor: RootCommandProcessor<SlashSender, S> = RootCommandProcessor(
            command,
            registryContainer,
            commandOptions
        )

        val name = processor.name

        // Get or add command, then add its sub commands
        val rootCommand: RootCommand<SlashSender, S> = guildCommands
            .getOrPut(guildId) { mutableMapOf() }
            .getOrPut(name) { RootCommand(processor) }

        rootCommand.addCommands(command, processor.commands(rootCommand))

        if (!isKordReady) {
            commandQueue.add { registerKordCommand(guildId, rootCommand) }
            return
        }

        kord.launch {
            registerKordCommand(guildId, rootCommand)
        }
    }

    override fun unregisterCommand(command: Any) {
        TODO("Not yet implemented")
    }

    override fun getRegistryContainer(): RegistryContainer<SlashSender, S> {
        return registryContainer
    }

    private suspend fun execute(event: ChatInputCommandInteractionCreateEvent) {
        val command = event.interaction.command

        val name = command.rootName

        val commands = ArrayDeque<String>()
        if (command is dev.kord.core.entity.interaction.SubCommand) {
            commands.add(command.name)
        }

        if (command is GroupCommand) {
            commands.add(command.groupName)
            commands.add(command.name)
        }

        val rootCommand = guildCommands[event.interaction.invokedCommandGuildId]?.get(name) ?: return

        val sender = commandOptions.commandExtensions.senderExtension.map(DummySender(event))

        // Mapping all arguments
        val arguments: Map<String, Pair<String, Any>> = command.options.map { (key, value) ->
            when (value) {
                is ResolvableOptionValue<*> -> {
                    val resolvedObject = value.resolvedObject
                    key to Pair(resolvedObject.toString(), resolvedObject as Any)
                }
                else -> {
                    val wrappedValue = value.value
                    key to Pair(wrappedValue.toString(), wrappedValue.toString() as Any)
                }
            }
        }.toMap()

        rootCommand.executeNonLinear(sender, null, commands, arguments)
    }

    private suspend fun suggest(event: GuildAutoCompleteInteractionCreateEvent) {
        event.interaction.suggestString {
            choice("hello", "test")
        }
    }

    private suspend fun registerKordCommand(guildId: Snowflake, rootCommand: RootCommand<SlashSender, S>) {
        kord.createGuildChatInputCommand(
            guildId,
            rootCommand.name,
            rootCommand.desc
        ) {
            // If only default then register with no groups or sub commands
            rootCommand.defaultCommand?.let {
                if (it is SubCommand<SlashSender, S>) {
                    options = it.mapArgumentsToKord()
                }
                return@createGuildChatInputCommand
            }

            val commands = rootCommand.commands.values

            commands.filterIsInstance<SubCommand<SlashSender, S>>().forEach {
                subCommand(it.name, it.desc) {
                    options = it.mapArgumentsToKord()
                }
            }

            commands.filterIsInstance<ParentSubCommand<SlashSender, S>>().forEach {
                group(it.name, it.desc) {
                    it.commands.values.filterIsInstance<SubCommand<SlashSender, S>>().forEach { sub ->
                        subCommand(sub.name, sub.desc) {
                            options = sub.mapArgumentsToKord()
                        }
                    }
                }
            }
        }
    }
}
